core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java

```diff
@@ -0,0 +1,626 @@
+      // field IDs, they must be determined before any writes occur
+      try {
+        return replaceTransaction();
+      } catch (NoSuchTableException e) {
+        return createTransaction();
+      }
+    }
+
+    private LoadTableResponse stageCreate() {
+      Map<String, String> tableProperties = propertiesBuilder.build();
+
+      CreateTableRequest request = CreateTableRequest.builder()
+          .stageCreate()
+          .withName(ident.name())
+          .withSchema(schema)
+          .withPartitionSpec(spec)
+          .withWriteOrder(writeOrder)
+          .withLocation(location)
+          .setProperties(tableProperties)
+          .build();
+
+      return client.post(
+          paths.tables(ident.namespace()), request, LoadTableResponse.class, headers(context),
+          ErrorHandlers.tableErrorHandler());
+    }
+  }
+
+  private static List<MetadataUpdate> createChanges(TableMetadata meta) {
+    ImmutableList.Builder<MetadataUpdate> changes = ImmutableList.builder();
+
+    Schema schema = meta.schema();
+    changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
+    changes.add(new MetadataUpdate.SetCurrentSchema(-1));
+
+    PartitionSpec spec = meta.spec();
+    if (spec != null && spec.isPartitioned()) {
+      changes.add(new MetadataUpdate.AddPartitionSpec(spec));
+      changes.add(new MetadataUpdate.SetDefaultPartitionSpec(-1));
+    }
+
+    SortOrder order = meta.sortOrder();
+    if (order != null && order.isSorted()) {
+      changes.add(new MetadataUpdate.AddSortOrder(order));
+      changes.add(new MetadataUpdate.SetDefaultSortOrder(-1));
+    }
+
+    String location = meta.location();
+    if (location != null) {
+      changes.add(new MetadataUpdate.SetLocation(location));
+    }
+
+    Map<String, String> properties = meta.properties();
+    if (properties != null && !properties.isEmpty()) {
+      changes.add(new MetadataUpdate.SetProperties(properties));
+    }
+
+    return changes.build();
+  }
+
+  private String fullTableName(TableIdentifier ident) {
+    return String.format("%s.%s", name(), ident);
+  }
+
+  private Map<String, String> fullConf(Map<String, String> config) {
+    Map<String, String> fullConf = Maps.newHashMap(properties());
+    fullConf.putAll(config);
+    return fullConf;
+  }
+
+  private Pair<RESTClient, FileIO> tableClients(Map<String, String> config) {
+    if (config.isEmpty()) {
+      return Pair.of(client, io); // reuse client and io since config is the same
+    }
+
+    Map<String, String> fullConf = fullConf(config);
+    String ioImpl = fullConf.get(CatalogProperties.FILE_IO_IMPL);
+    FileIO tableIO = CatalogUtil.loadFileIO(
+        ioImpl != null ? ioImpl : ResolvingFileIO.class.getName(), fullConf, this.conf);
+    RESTClient tableClient = clientBuilder.apply(fullConf);
+
+    return Pair.of(tableClient, tableIO);
+  }
+
+  /**
+   * Class to handle authorization headers and token refresh.
+   */
+  private class AuthSession {
```

Comment:
This class handles refreshing tokens (when `refresh()` is called) and maintains headers. Instances are passed as `Supplier<Map<String, String>>` from `session::headers` to the `HTTPClient`, which gets the header map just before each request.